File: bindings/java-jni/src/main/java/org/maplibre/nativejni/render/NativePointer.java
package org.maplibre.nativejni.render;

import java.util.Objects;

/**
 * Opaque borrowed native address value used for backend interop handles.
 *
 * <p>A {@code NativePointer} never owns the pointed-to resource and never grants general memory
 * access. It is only a carrier for backend-native handles accepted by render target descriptors or
 * returned by active texture frames.
 */
public final class NativePointer {
  public static final NativePointer NULL = new NativePointer(0, null);

  private final long address;
  private final FrameScope scope;

  private NativePointer(long address, FrameScope scope) {
    this.address = address;
    this.scope = scope;
  }

  /**
   * Borrows a backend-native address for interop with render target descriptors.
   *
   * <p>The caller keeps the referenced backend object alive and externally synchronized for the
   * full native borrow window documented by the descriptor or operation that receives this pointer.
   * A zero address returns {@link #NULL}.
   */
  public static NativePointer ofAddress(long address) {
    return address == 0 ? NULL : new NativePointer(address, null);
  }

  static NativePointer scoped(long address, FrameScope scope) {
    return address == 0 ? NULL : new NativePointer(address, Objects.requireNonNull(scope, "scope"));
  }

  public long address() {
    ensureActive();
    return address;
  }

  public boolean isNull() {
    ensureActive();
    return address == 0;
  }

  private void ensureActive() {
    if (scope != null) {
      scope.ensureActive();
    }
  }

  @Override
  public boolean equals(Object other) {
    return other instanceof NativePointer that && address() == that.address();
  }

  @Override
  public int hashCode() {
    return Long.hashCode(address());
  }

  @Override
  public String toString() {
    return "NativePointer[address=0x" + Long.toHexString(address()) + "]";
  }
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/render/NativePointer.kt
package org.maplibre.nativeffi.render

/** Opaque borrowed native address value used for backend interop handles. */
public class NativePointer
private constructor(private val addressValue: Long, private val scope: FrameScope?) {
  /** Borrowed address bit pattern. Access rejects use after a scoped frame closes. */
  public val address: Long
    get() {
      ensureActive()
      return addressValue
    }

  /** Returns true when this pointer represents a null backend handle. */
  public val isNull: Boolean
    get() {
      ensureActive()
      return addressValue == 0L
    }

  override fun equals(other: Any?): Boolean =
    other is NativePointer && addressValue == other.addressValue

  override fun hashCode(): Int = addressValue.hashCode()

  override fun toString(): String = "NativePointer[address=0x${addressValue.toString(16)}]"

  private fun ensureActive() {
    scope?.ensureActive()
  }

  public companion object {
    /** Null native pointer value. */
    public val NULL: NativePointer = NativePointer(0L, null)

    /** Creates an opaque borrowed pointer value from an address bit pattern. */
    public fun ofAddress(address: Long): NativePointer =
      if (address == 0L) NULL else NativePointer(address, null)

    internal fun scoped(address: Long, scope: FrameScope): NativePointer =
      if (address == 0L) NULL else NativePointer(address, scope)
  }
}
